File: backend/app/models/equipment_catalog.py
from sqlalchemy import Column, Integer, String, Float, DateTime, Text, Boolean, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from app.database import Base


class EquipmentCatalogItem(Base):
    __tablename__ = "equipment_catalog"

    id = Column(Integer, primary_key=True, index=True)
    type_code = Column(String(50), nullable=False, index=True)  # disconnector, breaker, zn, arrester, recloser, ...
    brand = Column(String(100), nullable=False, index=True)  # РЛНД, ВВ/TEL, ОПН, ...
    model = Column(String(120), nullable=False, index=True)  # РЛНД-10/400, ...
    full_name = Column(String(255), nullable=True)
    voltage_kv = Column(Float, nullable=True)
    current_a = Column(Float, nullable=True)
    manufacturer = Column(String(120), nullable=True)
    country = Column(String(80), nullable=True)
    description = Column(Text, nullable=True)
    attrs_json = Column(Text, nullable=True)  # произвольные характеристики JSON-строкой
    is_active = Column(Boolean, nullable=False, default=True, server_default="true")
    created_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), nullable=True)

    equipment = relationship("Equipment", back_populates="catalog_item")

File: frontend/lib/core/models/equipment_catalog.dart
class EquipmentCatalogItem {
  final int id;
  final String typeCode;
  final String brand;
  final String model;
  final String? fullName;
  final double? voltageKv;
  final double? currentA;
  final String? manufacturer;
  final String? country;
  final String? description;
  final String? attrsJson;
  final bool isActive;

  const EquipmentCatalogItem({
    required this.id,
    required this.typeCode,
    required this.brand,
    required this.model,
    this.fullName,
    this.voltageKv,
    this.currentA,
    this.manufacturer,
    this.country,
    this.description,
    this.attrsJson,
    required this.isActive,
  });

  static double? _toDouble(dynamic v) {
    if (v == null) return null;
    if (v is num) return v.toDouble();
    return double.tryParse(v.toString());
  }

  static int _toInt(dynamic v) {
    if (v is int) return v;
    if (v is num) return v.toInt();
    return int.tryParse(v.toString()) ?? 0;
  }

  factory EquipmentCatalogItem.fromJson(Map<String, dynamic> json) {
    return EquipmentCatalogItem(
      id: _toInt(json['id']),
      typeCode: (json['type_code'] ?? '').toString(),
      brand: (json['brand'] ?? '').toString(),
      model: (json['model'] ?? '').toString(),
      fullName: json['full_name']?.toString(),
      voltageKv: _toDouble(json['voltage_kv']),
      currentA: _toDouble(json['current_a']),
      manufacturer: json['manufacturer']?.toString(),
      country: json['country']?.toString(),
      description: json['description']?.toString(),
      attrsJson: json['attrs_json']?.toString(),
      isActive: json['is_active'] == true,
    );
  }
}

